File: trie/Design Add and Search Words Data Structure.go
type TrieNode struct {
    children [26]*TrieNode
    is_word bool
}

type WordDictionary struct {
    root *TrieNode
}


func Constructor() WordDictionary {
    return WordDictionary{
        root: &TrieNode{},
    }
}


func (this *WordDictionary) AddWord(word string)  {
    cur := this.root
    for i:= range word{
        if cur.children[word[i] - 'a'] == nil{
            cur.children[word[i] - 'a'] = &TrieNode{}          
        }
        cur = cur.children[word[i] - 'a']
    }
    cur.is_word = true
}


func (this *WordDictionary) Search(word string) bool {
    return this.helperSearch(word, this.root)
}


func (this *WordDictionary) helperSearch(word string, node *TrieNode) bool {
    if node == nil{
        return false
    }
    curr:= node
    for i:= range word{
        if word[i] == '.'{
            for j:=0;j<26;j++{
                if this.helperSearch(word[i+1:], curr.children[j]) {
                    return true
                }
            }
            return false
        } else {
            if curr.children[word[i]-'a'] == nil{
                return false
            }
            curr = curr.children[word[i]-'a']
        }
    }
    return curr.is_word
}

/**
 * Your WordDictionary object will be instantiated and called as such:
 * obj := Constructor();
 * obj.AddWord(word);
 * param_2 := obj.Search(word);
 */

File: trie/Design Add and Search Words Data Structure.py
class TrieNode():
    def __init__(self):
        self.children = {}
        self.word = False
    
class WordDictionary:
    def __init__(self):
        self.root = TrieNode()
        self.max_word_len = 0
        
    def addWord(self, word):
        cur = self.root
        count = 0
        
        for c in word:
            if c not in cur.children:
                cur.children[c] = TrieNode()
            cur = cur.children[c]
            count += 1
        self.max_word_len = max(count, self.max_word_len)
        cur.word = True
        
    def search(self, word):
        if len(word) > self.max_word_len:
            return False
        
        def dfs(j, root):
            cur = root
            for i in range(j, len(word)):
                c = word[i]
                if c == ".":
                    for child in cur.children.values():
                        if dfs(i+1, child):
                            return True
                    return False
                else:
                    if c not in cur.children:
                        return False
                    cur = cur.children[c]
            return cur.word
        return dfs(0, self.root)